### ISSUE TITLE
WalkingPad A1 Pro

### ISSUE THREAD

**Issue by Lazybone** [rytilahti/python-miio#797]
**Device information:**

  - Name(s) of the device: Xiaomi WalkingPad A1 Pro
  - Link:

Use `miiocli device --ip <ip address> --token <token>`.

  - Model: ksmb.walkingpad.v1
  - Hardware version: MTK7697
  - Firmware version: 2.0.5

**Additional context**

**Comment by thomasvs** [comment]
I've been looking for ways to get at data from my WalkingPad.  Have you been able to interface with it in any way?

**Comment by dmatora** [comment]
Dreaming about extracting steps data from the device

**Comment by syssi** [comment]
```
Some `get_prop` calls of the mihome app (react plugin):

get_prop ['mode', 'sp']
get_prop ["goal"]
get_prop ['initial']
get_prop ['offline']
get_prop ["lock"]
get_prop ["max"]
get_prop ['mode', 'sp', 'step', 'dist', 'time', 'button_id']
get_prop ["max", "sensitivity", "start_speed", "auto", "disp", "lock"]
get_prop ['mode', 'sp', 'step', 'time', 'dist', 'cal', 'goal']

Methods:

set_mode [mode] / set_mode[0] / set_mode [1] / set_mode [3]
set_offline [offline]
set_state ['run']
tutorial [index]
set_max [speed] / set_max [6]
set_lock [0]
set_sensitivity [index]
set_auto [index]
set_lock [index]
set_start_speed [speed]
set_disp [index]
set_speed [speed]
set_cali [1] / set_cali [0]
set_goal [typeIndex, goalValue]
set_insure [1]
```

Could somebody provide the responses of the `get_prop` calls and wants to play around with the methods? We need a description per method and possible values. I can try to provide some parameter types.

**Comment by dmatora** [comment]
`raw_command set_start_speed '["1.0"]'` finally made my WalkingPad autostart on speed 1! Hurray!

`raw_command get_prop '["mode"]'` returns
 - [0] while mode is auto
 - [1] while mode is manual
 - [2] when in standby
`raw_command get_prop '["step"]'` returns [2303]
`raw_command get_prop '["time"]'` returns [1970] (while time is 32:50)
`raw_command get_prop '["dist"]'` returns [1869]
`raw_command get_prop '["cal"]'` returns [67340]
`raw_command get_prop '["goal"]'` returns [0, 60]
`raw_command get_prop '["max"]'` returns [6.0]
`raw_command get_prop '["initial"]'` returns [3]
`raw_command get_prop '["offline"]'` returns [0]
`raw_command get_prop '["sensitivity"]'` returns [2]
`raw_command get_prop '["sp"]'` returns [1.0]
`raw_command get_prop '["start_speed"]'` returns [1.0]
`raw_command get_prop '["auto"]'` returns [1]
`raw_command get_prop '["disp"]'` returns [19]
`raw_command get_prop '["lock"]'` returns [0]

**Comment by dmatora** [comment]
Built myself an iOS app, to put steps from walking pad into apple health and watch my daily step goal
![telegram-cloud-photo-size-2-5244712315109093918-y](https://user-images.githubusercontent.com/647062/106666534-c9dca300-65b8-11eb-99cd-05e4f6eaa0f4.jpg)
![telegram-cloud-photo-size-2-5246693751026463818-y](https://user-images.githubusercontent.com/647062/106666585-d95bec00-65b8-11eb-8fd4-829b9d04491f.jpg)
unfortunately it only works when web server with php script is providing miiocli bridge on local network
hopefully milo javascript library will get react native support soon enough and I'll be able to make this standalone app

**Comment by dmatora** [comment]
> Some `get_prop` calls of the mihome app (react plugin):

@syssi btw, how did you get those?
I'd really like to get those for my Xiaomi Rice cooking robot (MBIH-R3), it's app is also a nightmare

**Comment by rytilahti** [comment]
@dmatora if you are talking about getting supported by https://github.com/aholstenson/miio , I wouldn't hold my breath as the last commit is from 2018. If you know of a maintained fork / other project, please let us know! I think it would be a good idea to have a section for "related projects" in the readme file.

**Comment by dmatora** [comment]
> feel free to create a PR to add support for the walking pad to this library, if you wish :-)
I am thinking about adding it to home assistant so it can act as a proxy for iOS app
does it involve adding it to milo python? or can I just add it straight to home assistant?

**Comment by rytilahti** [comment]
@dmatora yes, to get it supported in the official homeassistant, the support needs to be added first to this library and then to homeassistant (separately).

There is xiaomi_raw (linked in the README.md), which allows using devices without proper integration for the time being. 

If you are somewhat familiar with python, adding support to new device to this library is not usally very complicated, and there are plenty of examples for other devices available :-)

**Comment by dmatora** [comment]
can't remember writing anything in python
when getting token for walkingpad using `token_extractor.py` I've noticed it also shows ```
   NAME:  Thermometer
   ID:    blt.3.10dog0vhs5400
   IP:    185.135.151.12
   TOKEN: 1D1EAD2236B752B5B4B16250
   MODEL: miaomiaoce.thermo.t01
```
which is a bluetooth device.
Does miio_python work with bluetooth devices?
I would love to get Miaomiaoce Zenmeasure (or any other reliable continious body temperature monitor) integrated with Apple Health

**Comment by syssi** [comment]
The short answer is no. `python-miio` requires a wifi device (talking the miio protocol).

**Comment by rytilahti** [comment]
For BT(LE), https://github.com/zewelor/bt-mqtt-gateway looks nice, but I have never used it so far. There are some bluetooth standardized formats, but I suppose most use some sort of proprietary format and you have to manually reverse-engineer the payload contents to add support. If you want to go that road, see https://reverse-engineering-ble-devices.readthedocs.io/en/latest/ :)

**Comment by syssi** [comment]
You could grab a esp32 and use a esphome as btle gateway: https://esphome.io/components/sensor/xiaomi_ble.html#lywsd02

**Comment by dmatora** [comment]
I spoke to @Ernst79 and he added support for ZenMeasure to HomeAssistant, except that values sensor reports do not mach value shown on ZenMeasure app.
More details on issue here - https://github.com/custom-components/ble_monitor/issues/264

I'll think who can I speak to about including HomeAssistant support for Walkingpad :)

**Comment by dewgenenny** [comment]
Hey! I just added a pull request here: https://github.com/rytilahti/python-miio/pull/975 to support the walkingpad a1 :)

**Comment by dmatora** [comment]
> Hey! I just added a pull request here: https://github.com/rytilahti/python-miio/pull/975 to support the walkingpad a1 :) 

That's a great news! Thanks!
Hopefully it will be merged and we can start adding this to home assistant. This clearly needs more features in order to serve as server for GUI I wrote. I might be able to collaborate on that.

**Comment by dewgenenny** [comment]
> set_sensitivity

hey @syssi are you able to share where you got those get_prop commands from for the walkingpad? I am trying to hunt down if there is a 'total distance' value stored somewhere in the device and so far I've had no luck....

**Comment by syssi** [comment]
@dewgenenny If you use the Mi Home App and setup your miIO device a react.js based plugin is downloaded and used. I've extracted the used properties and methods of the Mi Home App plugin.

**Comment by dmatora** [comment]
@syssi how did you access react plugin?

**Comment by syssi** [comment]
@dmatora The archive is downloaded if you install/pair the device at your smartphone / mihome app. I've sent you the archive by mail (gmail.com).

**Comment by dmatora** [comment]
@syssi well we got walkingpad already figured out
I'm asking so that I can build better GUI for my Xiaomi Rice Cooking Robot MBIH-R3
It's plugin is horrible - it doesn't even update stat, unless you restart the plugin.
Do I need rooted device?
Which path of smartphone did you extract this from?

**Comment by syssi** [comment]
Could you create a new issue for the device? Do you know the model identifier (f.e. `ksmb.walkingpad.v1`) of the device?

**Comment by rytilahti** [comment]
This library is only useful for devices that use the miio/miot protocol over wifi. The communication with btle devices is usually done by manufacturer specific commands and are out of the scope for this library.

**Comment by Ernst79** [comment]
@thomasvs If you are sure that the non-pro version communicates over BLE (and you are using Home Assistant), we might be able to add it to [BLE monitor](https://github.com/custom-components/ble_monitor/) It's not supported yet, but if you create a HCIDUMP or use the `report_unknown` option of this integration, I can have a look if it is using the MiBeacon BLE protocol.

**Comment by dewgenenny** [comment]
I'm pretty sure mine is just the A1. Does yours have a button on the right
hand side? Try long pressing, then searching for a new wifi AP..... I
managed to set mine up using the xiaomi home app initially

**Comment by DorianRudolph** [comment]
I think there might be different versions? I have the "EU Vesion" of the A1 Pro and it does not seem to have WLAN support. The WalkingPad app uses bluetooth.

**Comment by DorianRudolph** [comment]
@thomasvs I can control my A1 Pro via BLE using the following code. It does not connect to the Xiaomi App, only to the WalkingPad app.

```py
import BLE_GATT
import time

pad_address = '57:4C:4E:29:07:F1'
att_uuid = '0000FE02-0000-1000-8000-00805F9B34FB'
start_cmd = bytes.fromhex("f7a20401a7fd")

sleep_cmd = bytes.fromhex("f7a20202a6fd")
manual_cmd = bytes.fromhex("f7a20201a5fd")
automatic_cmd = bytes.fromhex("f7a20200a4fd")


def set_speed(s):
    pad.char_write(att_uuid, (0xf7, 0xa2, 0x01, s, s + 0xa3, 0xfd))


def set_init_speed(s):
    pad.char_write(att_uuid, (0xf7, 0xa6, 0x04, 0x00, 0x00, 0x00, s, s + 0xaa, 0xfd))


pad = BLE_GATT.Central(pad_address)
pad.connect()

set_init_speed(20)
time.sleep(0.1)
pad.char_write(att_uuid, start_cmd)

pad.disconnect()
```

One can also read the status of the WalkingPad, which I didn't implement. At least mode and speed are obvious, when looking at the wireshark trace, but not sure how to receive the other statistics (maybe not even possible, because the app does not seem to display them correctly). Reverse engineering the WalkingPad app seems hard because it is written in flutter.

**Comment by dmatora** [comment]
I can not see any mention of `get_prop ["all"]` in the thread, which is a bummer
it provides response like this
```
['mode:2', 'time:0', 'sp:0.0', 'dist:0', 'cal:0', 'step:0']
```
and should be used to get these props, since each get_prop takes about second to complete, and commands can not be processed in parallel

**Comment by dewgenenny** [comment]
@dmatora you can check the code - there is a quick_status option - 

`    def quick_status(self) -> WalkingpadStatus:
        """Retrieve quick status.
        The walkingpad provides the option to retrieve a subset of properties in one call:
        steps, mode, speed, distance, calories and time.
        `status()` will do four more separate I/O requests for power, mode, start_speed, and sensitivity.
        If you don't need any of that, prefer this method for status updates.
        """
`

**Comment by rytilahti** [comment]
Maybe it would have been a good idea to have `quick_status()` -> `status()` and the slower variant as `full_status()` or something like that? :thinking:

**Comment by dmatora** [comment]
>Could you create a new issue for the device? Do you know the model identifier (f.e. ksmb.walkingpad.v1) of the device?

@syssi There is an existing discussion for Xiaomi Induction Cooker https://github.com/rytilahti/python-miio/pull/832

I left a comment there about chunmi.ihcooker.v2, can you help me with that one?

**Comment by Turkules** [comment]
> Built myself an iOS app, to put steps from walking pad into apple health and watch my daily step goal
> ![telegram-cloud-photo-size-2-5244712315109093918-y](https://user-images.githubusercontent.com/647062/106666534-c9dca300-65b8-11eb-99cd-05e4f6eaa0f4.jpg)
> ![telegram-cloud-photo-size-2-5246693751026463818-y](https://user-images.githubusercontent.com/647062/106666585-d95bec00-65b8-11eb-8fd4-829b9d04491f.jpg)
> unfortunately it only works when web server with php script is providing miiocli bridge on local network
> hopefully milo javascript library will get react native support soon enough and I'll be able to make this standalone app

Can you elaborate on your solution in more detail for "Non-experts" ?

**Comment by demirkolmustafa** [comment]
> Built myself an iOS app, to put steps from walking pad into apple health and watch my daily step goal
> ![telegram-cloud-photo-size-2-5244712315109093918-y](https://user-images.githubusercontent.com/647062/106666534-c9dca300-65b8-11eb-99cd-05e4f6eaa0f4.jpg)
> ![telegram-cloud-photo-size-2-5246693751026463818-y](https://user-images.githubusercontent.com/647062/106666585-d95bec00-65b8-11eb-8fd4-829b9d04491f.jpg)
> unfortunately it only works when web server with php script is providing miiocli bridge on local network
> hopefully milo javascript library will get react native support soon enough and I'll be able to make this standalone app

@dmatora I would happily pay for this app as this is a major painpoint for me with this product. I’ve reached out multiple times to Xiaomi on this but they won’t help, would be great if you can finish this app and launch it in the market

**Comment by dmatora** [comment]
@demirkolmustafa I just saw your message.
I've made a standalone app, that doesn't require php server about a year ago.
You can build it with Xcode and install on your iPhone.
Putting it on AppStore would cost me $100 per year and I don't think it will make any profit
